File: packages/shared-py/shared/gcs.py
"""GCS helpers used by Python stages (download source, upload outputs)."""

from __future__ import annotations

import os
from pathlib import Path

from google.cloud import storage as gcs_storage

from .env import required

_client: gcs_storage.Client | None = None


def storage() -> gcs_storage.Client:
    global _client
    if _client is None:
        _client = gcs_storage.Client()
    return _client


def bucket() -> gcs_storage.Bucket:
    return storage().bucket(required("GCS_BUCKET"))


def public_url(object_path: str) -> str:
    return f"https://storage.googleapis.com/{required('GCS_BUCKET')}/{object_path}"


def object_exists(object_path: str) -> bool:
    return bucket().blob(object_path).exists()


def download_file(object_path: str, local_path: str | Path) -> Path:
    local = Path(local_path)
    bucket().blob(object_path).download_to_filename(str(local))
    return local


def upload_file(object_path: str, local_path: str | Path, content_type: str = "application/octet-stream") -> str:
    blob = bucket().blob(object_path)
    blob.upload_from_filename(str(local_path), content_type=content_type)
    return public_url(object_path)

File: packages/shared-ts/src/gcs.ts
/** Shared GCS helpers used by API + compose + record-mix stages. */

import { Storage } from "@google-cloud/storage";
import { required } from "./env.js";

let _storage: Storage | undefined;

export function storage(): Storage {
  if (!_storage) _storage = new Storage();
  return _storage;
}

export function bucket() {
  return storage().bucket(required("GCS_BUCKET"));
}

/** V4 signed PUT URL for direct browser upload. 15-min default expiry. */
export async function signedPutUrl(
  objectPath: string,
  contentType: string,
  expiresInSeconds = 900,
): Promise<string> {
  const [url] = await bucket()
    .file(objectPath)
    .getSignedUrl({
      version: "v4",
      action: "write",
      expires: Date.now() + expiresInSeconds * 1000,
      contentType,
    });
  return url;
}

export function publicUrl(objectPath: string): string {
  const b = required("GCS_BUCKET");
  return `https://storage.googleapis.com/${b}/${objectPath}`;
}

export async function objectExists(objectPath: string): Promise<boolean> {
  const [exists] = await bucket().file(objectPath).exists();
  return exists;
}

export async function uploadFile(
  objectPath: string,
  localPath: string,
  contentType: string,
): Promise<string> {
  await bucket().upload(localPath, {
    destination: objectPath,
    metadata: { contentType },
  });
  return publicUrl(objectPath);
}

export async function downloadFile(objectPath: string, localPath: string): Promise<void> {
  await bucket().file(objectPath).download({ destination: localPath });
}
